packages/indy-sdk-to-askar-migration/src/IndySdkToAskarMigrationUpdater.ts

```diff
@@ -0,0 +1,416 @@
+      return
+    } else {
+      this.agent.config.logger.trace(`Moving '${this.backupFile}' back to the original path: '${this.dbPath}`)
+
+      // Move the backedup file back to the original path
+      await this.fs.copyFile(this.backupFile, this.dbPath)
+
+      this.agent.config.logger.trace(`Cleaned up the backed up file at '${this.backupFile}'`)
+    }
+  }
+
+  // Delete the backup as `this.fs.copyFile` only copies and no deletion
+  // Since we use `tempPath` which is cleared when certain events happen,
+  // e.g. cron-job and system restart (depending on the os) we could omit
+  // this call `await this.fs.delete(this.backupFile)`.
+  private async cleanBackup() {
+    this.agent.config.logger.trace(`Deleting the backup file at '${this.backupFile}'`)
+    await this.fs.delete(this.backupFile)
+  }
+
+  /**
+   * Move the migrated and updated database file to the new location according
+   * to the `FileSystem.dataPath`.
+   */
+  private async moveToNewLocation() {
+    const src = this.dbPath
+    // New path for the database
+    const dest = this.newWalletPath
+
+    // create the wallet directory
+    await this.fs.createDirectory(dest)
+
+    this.agent.config.logger.trace(`Moving upgraded database from ${src} to ${dest}`)
+
+    // Copy the file from the database path to the new location
+    await this.fs.copyFile(src, dest)
+
+    // Delete the original, only if specified by the user
+    if (this.deleteOnFinish) await this.fs.delete(this.dbPath)
+  }
+
+  /**
+   * Function that updates the values from an indy-sdk structure to the new askar structure.
+   *
+   * - Assert that the paths that will be used are free
+   * - Create a backup of the database
+   * - Migrate the database to askar structure
+   * - Update the Keys
+   * - Update the Master Secret (Link Secret)
+   * - Update the credentials
+   * If any of those failed:
+   *   - Revert the database
+   * - Clear the backup from the temporary directory
+   */
+  public async update() {
+    await this.assertDestinationsAreFree()
+
+    await this.backupDatabase()
+    try {
+      // Migrate the database
+      await this.migrate()
+
+      const keyMethod =
+        this.walletConfig?.keyDerivationMethod == KeyDerivationMethod.Raw ? StoreKeyMethod.Raw : StoreKeyMethod.Kdf
+      this.store = await Store.open({ uri: `sqlite://${this.dbPath}`, passKey: this.walletConfig.key, keyMethod })
+
+      // Update the values to reflect the new structure
+      await this.updateKeys()
+      await this.updateMasterSecret()
+      await this.updateCredentials()
+
+      // Move the migrated and updated file to the expected location for afj
+      await this.moveToNewLocation()
+    } catch (cause) {
+      this.agent.config.logger.error('Migration failed. Restoring state.')
+
+      await this.restoreDatabase()
+
+      throw new IndySdkToAskarMigrationError('Migration failed. State has been restored.', { cause })
+    } finally {
+      await this.cleanBackup()
```

Comment:
will finally be called if we throw in a catch?

Reply:
Yes, https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/try...catch#the_finally-block

> Immediately before a control-flow statement (return, throw, break, continue) is executed in the try block or catch block.